README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->

bayefdr
=======

<!-- badges: start -->

[![R build
status](https://github.com/VallejosGroup/bayefdr/workflows/R-CMD-check/badge.svg)](https://github.com/VallejosGroup/bayefdr/actions)
[![CRAN
status](https://www.r-pkg.org/badges/version/bayefdr)](https://CRAN.R-project.org/package=bayefdr)
[![downloads](http://cranlogs.r-pkg.org/badges/grand-total/bayefdr)](https://CRAN.R-project.org/package=bayefdr)
<!-- badges: end -->

The goal of bayefdr is to provide tools for the estimation and
optimisation of Bayesian expected false discovery and expected false
negative rates.

Installation
------------

You can install the released version of bayefdr from
[CRAN](https://CRAN.R-project.org) with:

    install.packages("bayefdr")
    # development version:
    ## devtools::install_github("VallejosGroup/bayefdr")

Usage
-----

The main functions in this package are `efdr`, `efnr` and `efdr_search`.
`efdr` and `efnr` calculate the EFDR or EFNR for a vector of
probabilities given a specified probability threshold. `efdr_search`
finds the probability threshold that matches a target EFDR as closely as
possible. The basic input to this function is a vector of probabilities
and a target EFDR.

    library("bayefdr")
    set.seed(42)
    probs <- runif(100)
    efdr(0.7, probs)
    #> [1] 0.1429126
    efnr(0.7, probs)
    #> [1] 0.3531349
    efdr <- efdr_search(probs, target_efdr = 0.1)
    efdr
    #> An object of class 'bayefdr'.
    #>  Optimal threshold: 0.768 EFDR: 0.0985 EFNR: 0.399

The output of this function is a `data.frame` with some extra
attributes. There is a plot method too.

    head(efdr)
    #>   threshold     EFDR      EFNR
    #> 1   0.50000 0.239581 0.2361073
    #> 2   0.50025 0.239581 0.2361073
    #> 3   0.50050 0.239581 0.2361073
    #> 4   0.50075 0.239581 0.2361073
    #> 5   0.50100 0.239581 0.2361073
    #> 6   0.50125 0.239581 0.2361073
    plot(efdr)

<img src="man/figures/README-outputs-1.png" width="100%" />


NEWS.md
# Version 0.1.0 (2020-07-30)

Initial CRAN release

# Version 0.2.0 (2021-09-29)

- Add cumplot and traceplot functionality


tests/testthat.R
library(testthat)
library(bayefdr)

test_check("bayefdr")


tests/testthat/test-plots.R
context("plots")

test_that("cumplot works", {
    expect_is(cumplot(rnorm(1000)), "gg")
})

test_that("cumplot works", {
    expect_is(traceplot(rnorm(1000)), "gg")
})


tests/testthat/test-bayefdr.R
context("bayefdr")

test_that("Basic functionality", {
    probs <- runif(100)
    efdr <- efdr_search(probs, target_efdr = 0.1)
    expect_is(efdr, "data.frame")
    expect_is(efdr, "bayefdr")
    expect_is(plot(efdr), "gg")
})


test_that("error conditions/arguments", {
    probs <- runif(100)
    expect_error(
        efdr_search(probs),
        "argument \"target_efdr\" is missing, with no default"
    )
    expect_error(
        efdr_search(probs, target_efdr = -1),
        "target_efdr not greater than or equal to 0"
    )
    expect_error(
        efdr_search(probs, target_efdr = 2),
        "target_efdr not less than 1"
    )
    expect_error(
        efdr_search(probs, target_efdr = 0.1, min_threshold = -1),
        "min_threshold not greater than 0.5"
    )
    expect_error(
        efdr_search(probs, target_efdr = 0.1, min_threshold = 2),
        "min_threshold not less than 1"
    )
})


test_that("optimal", {
    set.seed(42)
    probs <- runif(100)
    e <- efdr_search(probs, target_efdr = 0.1)
    expect_equal(round(e[optimal(e), "threshold"], digits = 3), 0.768)
    expect_equal(round(e[optimal(e), "EFDR"], digits = 3), 0.099)
    expect_equal(round(e[optimal(e), "EFNR"], digits = 3), 0.399)
})

test_that("NaN handling", {
    expect_error(efdr_search(rep(1, 10), target_efdr=0.1), NA)
    expect_error(efdr_search(c(0.8, 0.8, 1.0, 1.0, 0.8), target_efdr=0.1), NA)
})




R/cumplot.R
#' Plot the cumulative median, mean, and 95\% high posterior density region.
#' @param x An vector of MCMC draws.
#' @param ylab An optional y-axis label.
#' @param burn Integer specifying the number of initial iterations to be
#' discarded.
#' @param thin Integer specifying the thinning factor to be used on
#' the MCMC steps.
#' @param hpd_level Floating point specifying the desired HPD level.
#' @return A ggplot showing the cumulative mean, median and HPD.
#' @examples
#' x <- rnorm(1000)
#' cumplot(x)
#' @export
cumplot <- function(
        x,
        ylab = NULL,
        burn = 0,
        thin = 1,
        hpd_level = 0.95
    ) {
    x <- x[burn:length(x)]
    x <- x[seq(1, length(x), by = thin)]
    cum_median <- .cumulative_median(x)
    cum_HPD <- .cumulative_HPD(x, hpd_level)
    cum_mean <- .cumulative_mean(x)
    ggplot() +
        aes(x = seq_along(cum_median)) +
        geom_line(aes(y = cum_median, color = "Posterior median")) +
        geom_line(aes(y = cum_mean, color = "Posterior mean")) +
        geom_line(aes(y = cum_HPD[1, ], color = "HPD (lower)")) +
        geom_line(aes(y = cum_HPD[2, ], color = "HPD (upper)")) +
        scale_color_brewer(palette = "Paired", name = NULL) +
        labs(x = "Iteration", y = ylab)
}

.cumulative_median <- function(y) {
    n <- length(y)
    out <- vapply(
        seq_len(n)[-1],
        function(x) median(y[seq_len(x)]),
        numeric(1)
    )
    return(out)
}

.cumulative_HPD <- function(y, level) {
    n <- length(y)
    out <- vapply(
        seq_len(n)[-1],
        function(x) .HPDinterval(y[seq_len(x)], level),
        numeric(2)
    )
    return(out)
}

.cumulative_mean <- function(y) {
    (cumsum(y) / seq_along(y))[-1]
}


.HPDinterval <- function(obj, prob = 0.95) {
    obj <- as.matrix(obj)
    vals <- apply(obj, 2, sort)
    if (!is.matrix(vals)) {
        stop("obj must have nsamp > 1")
    }
    nsamp <- nrow(vals)
    npar <- ncol(vals)
    gap <- max(1, min(nsamp - 1, round(nsamp * prob)))
    init <- seq_len(nsamp - gap)
    inds <- apply(
        vals[init + gap, , drop = FALSE] - vals[init, , drop = FALSE],
        2,
        which.min
    )
    ans <- cbind(
        vals[cbind(inds, seq_len(npar))],
        vals[cbind(inds + gap, seq_len(npar))]
    )
    dimnames(ans) <- list(colnames(obj), c("lower", "upper"))
    attr(ans, "Probability") <- gap / nsamp
    ans
}


R/traceplot.R

#' Trace, marginal density histogram, and autocorrelation plot of MCMC draws.
#' @param x A vector of MCMC draws.
#' @param ylab An optional y-axis label.
#' @param log Logical scalar controlling whether the y-axis should be logged.
#' @return A plot created using \code{\link[cowplot]{plot_grid}} showing
#' the trace, marginal density histogram, and autocorrelation function of the
#' MCMC draws in \code{x}.
#' @examples
#' x <- rnorm(1000)
#' traceplot(x)
#' @export
traceplot <- function(
        x,
        ylab = NULL,
        log = FALSE
    ) {

    df1 <- data.frame(
        Iteration = seq_along(x),
        Draws = x
    )
    # Code inspired by https://tinyurl.com/vezqqud
    acf <- stats::acf(x, plot = FALSE)
    df2 <- data.frame(lag = acf$lag, acf = acf$acf)

    p1 <- ggplot(df1) +
        geom_point(
            aes_string(x = "Iteration", y = "Draws"),
            alpha = 0
        ) + 
        geom_line(aes_string(x = "Iteration", y = "Draws")) +
        scale_y_continuous(trans = if (log) "log10" else "identity") +
        labs(
            x = "Iteration",
            y = ylab
        )
    p1 <- ggExtra::ggMarginal(p1, type = "histogram", margins = "y")
    p2 <- ggplot(df2, aes_string(x = "lag", y = "acf")) +
        geom_hline(aes(yintercept = 0)) +
        geom_segment(mapping = aes_string(xend = "lag", yend = 0))
    cowplot::plot_grid(p1, p2)
}


R/bayefdr.R
#' Bayesian EFDR optimisation.
#' 
#' Given a vector of probabilities, this function finds the probability 
#' threshold that matches a target expected false discovery rate as closely
#' as possible.
#' 
#' @param probs Vector of probabilities.
#' @param min_threshold Minimum probability threshold. If the optimal 
#'  probability threshold is below this number, it is rejected and 
#'  \code{min_threshold} is used instead.
#' @param target_efdr Numeric scalar specifying the expected false discovery 
#'  rate to match.
#' @param prob_thresholds Vector for probability thresholds to scan,
#'  with the aim of finding the threshold that matches the target EFDR.
#' @return An object of class "bayefdr" containing the probability thresholds 
#'  tested, the EFDR and EFNR at each probability threshold, and the optimal 
#'  threshold.
#' @examples
#'  probs <- runif(100)
#'  efdr <- efdr_search(probs, target_efdr = 0.1)
#'  plot(efdr)
#' @export
efdr_search <- function(
        probs,
        target_efdr,
        min_threshold = 0.7,
        prob_thresholds = seq(0.5, 0.9995, by = 0.00025)
    ) {

    assert_that(
        target_efdr >= 0,
        target_efdr < 1,
        min_threshold > 0.5,
        min_threshold < 1,
        all(prob_thresholds >= 0.5),
        all(prob_thresholds < 1)
    )
    if (!min_threshold %in% prob_thresholds) {
        prob_thresholds <- sort(c(min_threshold, prob_thresholds))
    }

    efdr_grid <- scan(probs, type = "efdr", prob_thresholds = prob_thresholds)
    efnr_grid <- scan(probs, type = "efnr", prob_thresholds = prob_thresholds)

    abs_diff <- abs(efdr_grid - target_efdr)
    ind_opt <- which.min(abs_diff)
    efdr_opt <- efdr_grid[ind_opt]
    efnr_opt <- efnr_grid[ind_opt]
    optimal <- which(efdr_grid == efdr_opt & efnr_grid == efnr_opt)
    if (length(optimal) > 1) {
        optimal <- median(round(median(optimal)))
    }
    if (length(optimal) == 0) {
        message("EFDR estimation failed, returning specified min_threshold")
        return(
            bayefdr(
                which(prob_thresholds == min_threshold),
                prob_thresholds = prob_thresholds,
                efdr_grid = efdr_grid,
                efnr_grid = efnr_grid
            )
        )
    }
    if (prob_thresholds[optimal] < min_threshold) {
        ## issue warning and fix to input
        optimal <- which(prob_thresholds == min_threshold)
        message(
            "Unable to find a probability threshold that achieves the ",
            "desired EFDR +-0.025. ",
            "Returning specified min_threshold."
        )
    }
    optimal_threshold <- bayefdr(
        optimal,
        prob_thresholds = prob_thresholds,
        efdr_grid = efdr_grid,
        efnr_grid = efnr_grid
    )
}

#' Plot the EFDR, EFNR grids of a bayefdr object.
#' @param x An object of class bayefdr.
#' @param ... Unused.
#' @return A ggplot.
#' @export
plot.bayefdr <- function(x, ...) {
    mdf <- melt(x, measure.vars = c("EFDR", "EFNR"))
    ggplot(mdf, aes_string(x = "threshold", y = "value", colour = "variable")) +
        geom_line(na.rm = TRUE) +
        labs(x = "Probability threshold", y = "Error rate") +
        ylim(0:1) +
        scale_colour_brewer(name = "", palette = "Set2") +
        geom_hline(
            aes(
                yintercept = x[optimal(x), "EFDR"],
                colour = "Selected EFDR"
            ),
            linetype = 2,
            na.rm = TRUE
        ) +
        geom_vline(
            aes(xintercept = x[optimal(x), "threshold"],
                colour = "Selected threshold"
            )
        )
}

#' EFDR and EFNR estimation
#' 
#' Calculate the Expected False Discovery Rate (EFDR) or Expected False 
#' Negative Rate (EFNR) in a vector of probabilities, given a specified
#' evidence threshold.
#' @param evidence_threshold Scalar value specifying the evidence threshold
#' at which the EFDR or EFNR should be evaluated.
#' @param probs Vector of probabilities.
#' @rdname efdr-efnr
#' @export
efdr <- function(evidence_threshold, probs) {
    sum((1 - probs) * (probs > evidence_threshold)) /
        sum(probs > evidence_threshold)
}
#' @rdname efdr-efnr
#' @export
efnr <- function(evidence_threshold, probs) {
    sum(probs * (evidence_threshold >= probs)) /
        sum(evidence_threshold >= probs)
}

scan <- function(
        probs,
        type = c("efdr", "efnr"),
        prob_thresholds
    ) {
    type <- match.arg(type)
    fun <- match.arg(type)
    grid <- vapply(
        prob_thresholds,
        FUN = fun,
        FUN.VALUE = numeric(1),
        probs = probs
    )
}

bayefdr <- function(optimal, prob_thresholds, efdr_grid, efnr_grid) {
    structure(
        data.frame(
            threshold = prob_thresholds,
            EFDR = efdr_grid,
            EFNR = efnr_grid
        ),
        optimal = optimal,
        class = c("bayefdr", "data.frame")
    )
}

#' Retrieve the index of the optimal probability threshold.
#' @param x An object of class "bayefdr".
#' @return The integer index of the optimal probability threshold.
#' @examples
#'  probs <- runif(100)
#'  e <- efdr_search(probs, target_efdr = 0.1)
#'  optimal(e)
#'  e[optimal(e), ]
#' @export
optimal <- function(x) attr(x, "optimal")

#' Print methods for bayefdr objects.
#' @param x An object of class bayefdr.
#' @param ... Unused.
#' @rdname utils
#' @export
print.bayefdr <- function(x, ...) {
    df <- format(x[optimal(x), ], digits = 3)
    cat(
        "An object of class 'bayefdr'.\n",
        "Optimal threshold:", df$threshold, "EFDR:", df$EFDR, "EFNR:", df$EFNR, 
        "\n"
    )
}

#' @rdname utils
#' @export
head.bayefdr <- function(x, ...) {
    head(as.data.frame(x))
}
